cpp/oneapi/dal/algo/decision_forest/backend/gpu/feature_type_helper_dpc.cpp

```diff
@@ -0,0 +1,329 @@
+    size_t local_blocks_count,
+    const dal::backend::event_vector& deps) {
+    ONEDAL_ASSERT(values_nd.get_count() == row_count_);
+    ONEDAL_ASSERT(indices_nd.get_count() == row_count_);
+    ONEDAL_ASSERT(bin_borders_nd.get_count() == max_bins_);
+    ONEDAL_ASSERT(bins_nd.get_count() == row_count_);
+
+    const sycl::nd_range<1> nd_range =
+        be::make_multiple_nd_range_1d(de::check_mul_overflow(local_size, local_blocks_count),
+                                      local_size);
+
+    const Index row_count = row_count_;
+    const Float* values = values_nd.get_data();
+    const Index* indices = indices_nd.get_data();
+    const Float* bin_borders = bin_borders_nd.get_data();
+    Bin* bins = bins_nd.get_mutable_data();
+
+    auto event = queue_.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
+            auto sbg = item.get_sub_group();
+            const std::uint32_t n_groups = item.get_group_range(0);
+            const std::uint32_t n_sub_groups = sbg.get_group_range()[0];
+            const std::uint32_t n_total_sub_groups = n_sub_groups * n_groups;
+            const Index elems_for_sbg =
+                row_count / n_total_sub_groups + bool(row_count % n_total_sub_groups);
+            const std::uint32_t local_size = sbg.get_local_range()[0];
+
+            const std::uint32_t local_id = sbg.get_local_id();
+            const std::uint32_t sub_group_id = sbg.get_group_id();
+            const std::uint32_t group_id = item.get_group().get_id(0) * n_sub_groups + sub_group_id;
+
+            Index ind_start = group_id * elems_for_sbg;
+            Index ind_end =
+                sycl::min(static_cast<Index>((group_id + 1) * elems_for_sbg), row_count);
+
+            Index cur_bin = 0;
+
+            for (Index i = ind_start + local_id; i < ind_end; i += local_size) {
+                Float value = values[i];
+                while (bin_borders[cur_bin] < value)
+                    cur_bin++;
+                bins[indices[i]] = cur_bin;
+            }
+        });
+    });
+
+    return event;
+}
+
+template <typename Float, typename Bin, typename Index>
+sycl::event indexed_features<Float, Bin, Index>::compute_bins(
+    const pr::ndarray<Float, 1>& values_nd,
+    const pr::ndarray<Index, 1>& indices_nd,
+    pr::ndarray<Bin, 1>& bins_nd,
+    feature_entry& entry,
+    const dal::backend::event_vector& deps) {
+    ONEDAL_ASSERT(values_nd.get_count() == row_count_);
+    ONEDAL_ASSERT(indices_nd.get_count() == row_count_);
+    ONEDAL_ASSERT(bins_nd.get_count() == row_count_);
+
+    const Index max_bins = std::min(max_bins_, row_count_);
+    const Index local_size = preferable_sbg_size_;
+    const Index local_block_count = max_local_block_count_ * local_size < row_count_
+                                        ? max_local_block_count_
+                                        : (row_count_ / local_size) + bool(row_count_ % local_size);
```

Comment:
I think, it's time to create common function. I saw expressions for block size computations multiple times in various DPC++ kernels.